File: NWorldPermissions/src/top/nololiyt/worldpermissions/commands/Executor.java
package top.nololiyt.worldpermissions.commands;

import org.bukkit.command.CommandSender;
import top.nololiyt.worldpermissions.RootPlugin;
import top.nololiyt.worldpermissions.entitiesandtools.DotDividedStringBuilder;
import top.nololiyt.worldpermissions.entitiesandtools.StringPair;

import java.util.List;
import java.util.Set;


public abstract class Executor implements CommandLayer
{
    protected abstract boolean run(int layer,
                                   RootPlugin rootPlugin,
                                   DotDividedStringBuilder permission,
                                   DotDividedStringBuilder messageKey,
                                   CommandSender commandSender,
                                   String[] args);
    
    public void execute(int layer,
                        RootPlugin rootPlugin,
                        DotDividedStringBuilder permission,
                        DotDividedStringBuilder messageKey,
                        CommandSender commandSender,
                        String[] args)
    {
        if (permissionName() != null)
        {
            permission.append(permissionName());
            if (!commandSender.hasPermission(permission.toString()))
            {
                return;
            }
        }
        DotDividedStringBuilder messageKeyCopy = new DotDividedStringBuilder(messageKey);
        
        if (messageKey() != null)
            messageKey.append(messageKey());
    
        if (!run(layer, rootPlugin, permission, messageKey, commandSender, args))
        {
            sendHelp(messageKeyCopy, rootPlugin, commandSender);
        }
    }
    
    protected boolean sendHelp(DotDividedStringBuilder messageKey,
                               RootPlugin rootPlugin, CommandSender commandSender)
    {
        messageKey.append("help");
        StringPair[] pairs = new StringPair[]{
                StringPair.senderName(commandSender.getName())
        };
    
        rootPlugin.getMessagesManager().sendMessage(commandSender, messageKey, pairs);
        return true;
    }
}

File: NWorldPermissions/src/top/nololiyt/worldpermissions/commands/Executor.kt
package top.nololiyt.worldpermissions.commands

import org.bukkit.command.CommandSender
import top.nololiyt.worldpermissions.RootPlugin
import top.nololiyt.worldpermissions.entities.DotDividedStringBuilder
import top.nololiyt.worldpermissions.entities.StringPair


abstract class Executor : CommandLayer
{
    protected abstract fun permissionName(): String?
    protected abstract fun messageKey(): String?
    
    protected abstract fun run(layer: Int, rootPlugin: RootPlugin, permission: DotDividedStringBuilder,
        messageKey: DotDividedStringBuilder, commandSender: CommandSender, args: Array<String>): Boolean
    
    override fun execute(layer: Int, rootPlugin: RootPlugin, permission: DotDividedStringBuilder,
        messageKey: DotDividedStringBuilder, commandSender: CommandSender, args: Array<String>)
    {
        if (permissionName() != null)
        {
            permission.append(permissionName());
            if (!commandSender.hasPermission(permission.toString())) return;
        }
        val messageKeyCopy = DotDividedStringBuilder(messageKey);
    
        if (messageKey() != null) messageKey.append(messageKey());
    
        if (!run(layer, rootPlugin, permission, messageKey, commandSender, args))
        {
            sendHelp(messageKeyCopy, rootPlugin, commandSender);
        }
    }
    
    private fun sendHelp(messageKey: DotDividedStringBuilder, rootPlugin: RootPlugin,
        commandSender: CommandSender): Boolean
    {
        messageKey.append("help");
        val pairs = arrayOf<StringPair?>(StringPair.senderName(commandSender.name));
        
        rootPlugin.messagesManager.sendMessage(messageKey, pairs, commandSender);
        return true;
    }
}
